File: rcc/gui/game_list.py
from PyQt5.QtWidgets import (
    QDialog,
    QVBoxLayout,
    QTableWidget,
    QTableWidgetItem,
    QScrollArea,
    QAbstractScrollArea,
    QHeaderView,
    QWidget,
    QAbstractItemView,
    QLineEdit,
)
from PyQt5.QtCore import Qt
from PyQt5.QtGui import QIcon, QPixmap, QColor

from rcc.models.mangohud_level import MangoHudLevel
from rcc.models.wine_sync_option import WineSyncOption
from rcc.services.hardware_service import HARDWARE_SERVICE
from rcc.services.steam_service import STEAM_SERVICE
from rcc.utils.constants import ICONS_PATH
from rcc.utils.events import STEAM_SERVICE_CONNECTED
from rcc.utils.gui_utils import NoScrollComboBox
from rcc.utils.beans import EVENT_BUS, TRANSLATOR


class GameList(QDialog):
    """Window for game list"""

    INSTANCE: QDialog = None

    def __init__(
        self, parent: QWidget, manage_parent=False, app_id: int = None
    ):  # pylint: disable=too-many-locals, too-many-branches, too-many-statements
        super().__init__(parent)
        if GameList.INSTANCE is None:  # pylint: disable=too-many-nested-blocks
            GameList.INSTANCE = self
            self.__parent = parent
            self.__manage_parent = manage_parent
            self.setWindowTitle(TRANSLATOR.translate("game.performance.configuration"))
            self.setFixedSize(1000, 600)
            self.setWindowIcon(QIcon(f"{ICONS_PATH}/icon-45x45.png"))
            self.setAttribute(Qt.WA_DeleteOnClose)
            self.setWindowFlags(self.windowFlags() & ~Qt.WindowContextHelpButtonHint)

            # Configurar layout principal
            layout = QVBoxLayout(self)

            # Crear el área de scroll
            scroll_area = QScrollArea(self)
            scroll_area.setWidgetResizable(True)
            scroll_area.setVerticalScrollBarPolicy(Qt.ScrollBarAsNeeded)
            scroll_area.setHorizontalScrollBarPolicy(Qt.ScrollBarAsNeeded)

            # Crear un widget contenedor
            container = QWidget()
            scroll_area.setWidget(container)

            # Layout del contenedor
            container_layout = QVBoxLayout(container)

            # Crear la tabla
            game_cfg = STEAM_SERVICE.get_games()
            appids = list(game_cfg.keys())
            if app_id is not None:
                appids.append(app_id)

            columns = [
                TRANSLATOR.translate("game.title"),
                TRANSLATOR.translate("used.gpu"),
                TRANSLATOR.translate("used.steamdeck"),
            ]

            if STEAM_SERVICE.metrics_enabled:
                columns.append(TRANSLATOR.translate("metrics"))

            columns.append(TRANSLATOR.translate("winesync"))
            columns.append(TRANSLATOR.translate("environment"))
            columns.append(TRANSLATOR.translate("params"))

            table = QTableWidget(len(appids), len(columns))
            table.setSizeAdjustPolicy(QAbstractScrollArea.AdjustToContents)
            table.setHorizontalHeaderLabels(columns)
            table.verticalHeader().setVisible(False)

            # Configurar el ancho de las columnas
            header = table.horizontalHeader()
            # Primera columna se expande
            header.setSectionResizeMode(0, QHeaderView.Stretch)
            # Resto de columnas ajusta su tamaño al contenido
            for i in range(1, len(columns)):
                header.setSectionResizeMode(i, QHeaderView.ResizeToContents)
            header.setHighlightSections(False)  # Evitar que el encabezado se resalte

            # Desactivar edición y selección en la tabla
            table.setEditTriggers(QAbstractItemView.NoEditTriggers)
            table.setSelectionMode(QAbstractItemView.NoSelection)
            table.setFocusPolicy(Qt.NoFocus)

            row = 0
            for appid in sorted(appids, key=lambda e: game_cfg[e].name):
                is_running = appid in STEAM_SERVICE.running_games
                try:
                    col = 0
                    # Title

                    item = QTableWidgetItem(
                        game_cfg[appid].name
                        if not is_running
                        else game_cfg[appid].name + " (" + TRANSLATOR.translate("running") + "...)"
                    )
                    if STEAM_SERVICE.get_icon_path(appid):
                        pixmap = QPixmap(STEAM_SERVICE.get_icon_path(appid)).scaled(
                            32, 32, Qt.KeepAspectRatio, Qt.SmoothTransformation
                        )
                        icon = QIcon(pixmap)
                        item.setIcon(icon)
                    item.setFlags(Qt.ItemIsEnabled)
                    item.setToolTip(f"{appid}")
                    if is_running:
                        item.setForeground(QColor("green"))
                    table.setItem(row, col, item)
                    col += 1

                    # GPU
                    gpu_combo = NoScrollComboBox()
                    gpu_combo.setEnabled(not is_running)  # Usar la subclase personalizada
                    gpu_combo.addItem(TRANSLATOR.translate("label.dgpu.auto"), None)
                    gpu_combo.setCurrentIndex(0)
                    for i, gpu in enumerate(HARDWARE_SERVICE.gpus):
                        gpu_combo.addItem(f"{gpu.value.capitalize()}", gpu)
                        if gpu == STEAM_SERVICE.get_prefered_gpu(appid):
                            gpu_combo.setCurrentIndex(i + 1)

                    gpu_combo.currentIndexChanged.connect(
                        lambda _, widget=gpu_combo, game=appid: self.__on_gpu_changed(  # pylint: disable=line-too-long
                            widget, game
                        )
                    )
                    table.setCellWidget(row, col, gpu_combo)
                    col += 1

                    # GPU
                    steamdeck_combo = NoScrollComboBox()
                    steamdeck_combo.setEnabled(not is_running)  # Usar la subclase personalizada
                    steamdeck_combo.addItem(TRANSLATOR.translate("label.steamdeck.no"), False)
                    steamdeck_combo.addItem(TRANSLATOR.translate("label.steamdeck.yes"), True)
                    steamdeck_combo.setCurrentIndex(1 if STEAM_SERVICE.is_steamdeck(appid) else 0)

                    steamdeck_combo.currentIndexChanged.connect(
                        lambda _, widget=steamdeck_combo, game=appid: self.__on_steamdeck_changed(  # pylint: disable=line-too-long
                            widget, game
                        )
                    )
                    table.setCellWidget(row, col, steamdeck_combo)
                    col += 1

                    # Metrics
                    if STEAM_SERVICE.metrics_enabled:
                        metrics_combo = NoScrollComboBox()  # Usar la subclase personalizada
                        metrics_combo.setEnabled(not is_running)
                        for level in MangoHudLevel:
                            metrics_combo.addItem(TRANSLATOR.translate(f"label.level.{level}"), level)
                            if level == STEAM_SERVICE.get_metrics_level(appid):
                                metrics_combo.setCurrentIndex(level.value)

                        metrics_combo.currentIndexChanged.connect(
                            lambda _, widget=metrics_combo, game=appid: self.__on_metrics_changed(  # pylint: disable=line-too-long
                                widget, game
                            )
                        )
                        table.setCellWidget(row, col, metrics_combo)
                        col += 1

                    sync_combo = NoScrollComboBox()  # Usar la subclase personalizada
                    for i, option in enumerate(WineSyncOption):
                        sync_combo.addItem(TRANSLATOR.translate(f"label.winesync.{option}"), option)
                        if option == STEAM_SERVICE.get_wine_sync(appid):
                            sync_combo.setCurrentIndex(i)

                    sync_combo.currentIndexChanged.connect(
                        lambda _, widget=sync_combo, game=appid: self.__on_winesync_changed(  # pylint: disable=line-too-long
                            widget, game
                        )
                    )
                    sync_combo.setEnabled(not is_running and STEAM_SERVICE.is_proton(appid))
                    table.setCellWidget(row, col, sync_combo)
                    col += 1

                    env_input = QLineEdit()
                    env_input.setText(STEAM_SERVICE.get_environment(appid))
                    env_input.setEnabled(not is_running)
                    env_input.textChanged.connect(lambda text, game=appid: self.__on_environment_changed(text, game))
                    table.setCellWidget(row, col, env_input)
                    col += 1

                    args_input = QLineEdit()
                    args_input.setText(STEAM_SERVICE.get_parameters(appid))
                    args_input.setEnabled(not is_running)
                    args_input.textChanged.connect(lambda text, game=appid: self.__on_params_changed(text, game))
                    table.setCellWidget(row, col, args_input)
                    col += 1

                    row += 1
                except Exception as e:
                    print(f"{e}")

            # Añadir la tabla al layout del contenedor
            container_layout.addWidget(table)

            # Añadir el área de scroll al layout principal
            layout.addWidget(scroll_area)

            EVENT_BUS.on(STEAM_SERVICE_CONNECTED, self.close)

    def __on_steamdeck_changed(self, widget, game: int):
        sd = widget.currentData()
        STEAM_SERVICE.set_steamdeck(sd, game)

    def __on_params_changed(self, text, game: int):
        STEAM_SERVICE.set_parameters(text, game)

    def __on_environment_changed(self, text, game: int):
        STEAM_SERVICE.set_environment(text, game)

    def __on_metrics_changed(self, widget, game: int):
        level = widget.currentData()
        STEAM_SERVICE.set_metrics_level(level, game)

    def __on_winesync_changed(self, widget, game: int):
        level = widget.currentData()
        STEAM_SERVICE.set_wine_sync(level, game)

    def __on_gpu_changed(self, widget, game: int):
        gpu_brand = widget.currentData()
        STEAM_SERVICE.set_prefered_gpu(gpu_brand, game)

    def show(self):
        """Override default show behaviour"""
        if GameList.INSTANCE == self:
            if self.__manage_parent:
                self.__parent.show()
            super().show()
            if self.__manage_parent:
                self.__parent.hide()

    def closeEvent(self, event):  # pylint: disable=invalid-name
        """Override default closeEvent behaviour"""
        if GameList.INSTANCE == self:
            if self.__manage_parent:
                self.__parent.closeEvent(event)
            super().closeEvent(event)
            GameList.INSTANCE = None

File: RccCore/src/gui/game_list.cpp
#include "../../include/gui/game_list.hpp"

#include <QCloseEvent>
#include <QColor>
#include <QComboBox>
#include <QDialog>
#include <QHeaderView>
#include <QIcon>
#include <QLineEdit>
#include <QPixmap>
#include <QScrollArea>
#include <QString>
#include <QTableWidget>
#include <QVBoxLayout>
#include <QWidget>

#include "../../include/models/hardware/gpu_brand.hpp"
#include "../../include/models/steam/mangohud_level.hpp"
#include "../../include/models/steam/wine_sync_option.hpp"
#include "../../include/services/hardware_service.hpp"
#include "../../include/services/steam_service.hpp"	 // tu clase equivalente
#include "../../include/translator/translator.hpp"
#include "../../include/utils/gui_utils.hpp"
#include "RccCommons.hpp"

GameList::GameList(QWidget* parent, bool manage_parent) : QDialog(parent), manageParent(manage_parent), parentWidget(parent) {
	if (!INSTANCE)
		INSTANCE = this;

	setWindowTitle(QString::fromStdString(translator.translate("game.performance.configuration")));
	setFixedSize(1000, 600);
	setWindowIcon(QIcon(QString::fromStdString(Constants::ICON_45_FILE)));
	setAttribute(Qt::WA_DeleteOnClose);
	setWindowFlags(windowFlags() & ~Qt::WindowContextHelpButtonHint);

	QVBoxLayout* layout = new QVBoxLayout(this);

	QScrollArea* scrollArea = new QScrollArea(this);
	scrollArea->setWidgetResizable(true);
	scrollArea->setVerticalScrollBarPolicy(Qt::ScrollBarAsNeeded);
	scrollArea->setHorizontalScrollBarPolicy(Qt::ScrollBarAsNeeded);

	QWidget* container = new QWidget();
	scrollArea->setWidget(container);

	QVBoxLayout* containerLayout = new QVBoxLayout(container);

	auto gameCfg = steamService.getGames();
	QVector<unsigned int> appIds;
	for (const auto& [key, val] : gameCfg)
		appIds.append(static_cast<unsigned int>(std::stoul(key)));

	auto metricsEnabled = steamService.metricsEnabled();

	QStringList columns;
	columns << QString::fromStdString(translator.translate("game.title")) << QString::fromStdString(translator.translate("used.gpu"))
			<< QString::fromStdString(translator.translate("used.steamdeck"));
	if (metricsEnabled)
		columns << QString::fromStdString(translator.translate("metrics"));
	columns << QString::fromStdString(translator.translate("winesync")) << QString::fromStdString(translator.translate("environment"))
			<< QString::fromStdString(translator.translate("params"));

	QTableWidget* table = new QTableWidget(appIds.size(), columns.size());
	table->setSizeAdjustPolicy(QAbstractScrollArea::AdjustToContents);
	table->setHorizontalHeaderLabels(columns);
	table->verticalHeader()->setVisible(false);

	QHeaderView* header = table->horizontalHeader();
	header->setSectionResizeMode(0, QHeaderView::Stretch);
	for (int i = 1; i < columns.size(); ++i)
		header->setSectionResizeMode(i, QHeaderView::ResizeToContents);
	header->setHighlightSections(false);

	table->setEditTriggers(QAbstractItemView::NoEditTriggers);
	table->setSelectionMode(QAbstractItemView::NoSelection);
	table->setFocusPolicy(Qt::NoFocus);

	int row = 0;
	std::sort(appIds.begin(), appIds.end(), [&](int a, int b) { return gameCfg[std::to_string(a)].name > gameCfg[std::to_string(b)].name; });

	for (unsigned int appid : appIds) {
		bool isRunning = steamService.isRunning(appid);

		// --- Title ---
		QTableWidgetItem* item = new QTableWidgetItem(
			QString::fromStdString(gameCfg[std::to_string(appid)].name + (isRunning ? " (" + translator.translate("running") + "...)" : "")));

		item->setFlags(Qt::ItemIsEnabled);
		item->setToolTip(QString::number(appid));
		if (isRunning)
			item->setForeground(QColor("green"));

		table->setItem(row, 0, item);
		// --- GPU ---
		NoScrollComboBox* gpuCombo = new NoScrollComboBox();
		gpuCombo->setEnabled(!isRunning);
		gpuCombo->addItem(QString::fromStdString(translator.translate("label.dgpu.auto")), QString::fromStdString(""));
		int gpuIndex = 0;
		auto gpus	 = hardwareService.getGpus();
		int i		 = 0;
		for (const auto& [key, val] : gpus) {
			gpuCombo->addItem(QString::fromStdString(StringUtils::capitalize(key)), QString::fromStdString(key));
			if (steamService.getPreferedGpu(appid).has_value() && key == steamService.getPreferedGpu(appid).value().toString())
				gpuIndex = i + 1;
			i++;
		}
		gpuCombo->setCurrentIndex(gpuIndex);

		connect(gpuCombo, QOverload<int>::of(&QComboBox::currentIndexChanged), [=](int) {
			std::string str				  = gpuCombo->currentData().toString().toStdString();
			std::optional<GpuBrand> brand = std::nullopt;
			if (!str.empty())
				brand = GpuBrand::fromString(str);

			steamService.setPreferedGpu(appid, brand);
		});
		table->setCellWidget(row, 1, gpuCombo);

		// --- Steamdeck ---
		NoScrollComboBox* sdCombo = new NoScrollComboBox();
		sdCombo->setEnabled(!isRunning);
		sdCombo->addItem(QString::fromStdString(translator.translate("label.steamdeck.no")), false);
		sdCombo->addItem(QString::fromStdString(translator.translate("label.steamdeck.yes")), true);
		sdCombo->setCurrentIndex(steamService.isSteamDeck(appid) ? 1 : 0);

		connect(sdCombo, QOverload<int>::of(&QComboBox::currentIndexChanged),
				[=](int) { steamService.setSteamDeck(appid, sdCombo->currentData().toBool()); });
		table->setCellWidget(row, 2, sdCombo);

		// --- Metrics ---
		int col = 3;
		if (steamService.metricsEnabled()) {
			NoScrollComboBox* metricsCombo = new NoScrollComboBox();
			metricsCombo->setEnabled(!isRunning);
			auto items = MangoHudLevel::getAll();
			for (MangoHudLevel level : items) {
				metricsCombo->addItem(QString::fromStdString(translator.translate("label.level." + std::to_string(level.toInt()))), level.toInt());
				if (level == steamService.getMetricsLevel(appid))
					metricsCombo->setCurrentIndex(level.toInt());
			}

			connect(metricsCombo, QOverload<int>::of(&QComboBox::currentIndexChanged),
					[=](int) { steamService.setMetricsLevel(appid, MangoHudLevel::fromInt(metricsCombo->currentData().toInt())); });
			table->setCellWidget(row, col++, metricsCombo);
		}

		// --- WineSync ---
		NoScrollComboBox* syncCombo = new NoScrollComboBox();
		auto items					= WineSyncOption::getAll();
		i							= 0;
		for (WineSyncOption opt : items) {
			syncCombo->addItem(QString::fromStdString(translator.translate("label.winesync." + opt.toString())),
							   QString::fromStdString(opt.toString()));
			if (opt == steamService.getWineSync(appid))
				syncCombo->setCurrentIndex(i);
			i++;
		}
		syncCombo->setEnabled(!isRunning && steamService.isProton(appid));
		connect(syncCombo, QOverload<int>::of(&QComboBox::currentIndexChanged),
				[=](int) { steamService.setWineSync(appid, syncCombo->currentData().toString().toStdString()); });
		table->setCellWidget(row, col++, syncCombo);

		// --- Environment ---
		QLineEdit* envInput = new QLineEdit(QString::fromStdString(steamService.getEnvironment(appid)));
		envInput->setEnabled(!isRunning);
		connect(envInput, &QLineEdit::textChanged, [=](const QString& text) { steamService.setEnvironment(appid, text.toStdString()); });
		table->setCellWidget(row, col++, envInput);

		// --- Parameters ---
		QLineEdit* paramsInput = new QLineEdit(QString::fromStdString(steamService.getParameters(appid)));
		paramsInput->setEnabled(!isRunning);
		connect(paramsInput, &QLineEdit::textChanged, [=](const QString& text) { steamService.setParameters(appid, text.toStdString()); });
		table->setCellWidget(row, col++, paramsInput);

		row++;
	}

	containerLayout->addWidget(table);
	layout->addWidget(scrollArea);
}

void GameList::closeEvent(QCloseEvent* event) {
	if (INSTANCE == this)
		INSTANCE = nullptr;
	QDialog::closeEvent(event);
}